src/_P021_Level.ino

```diff
@@ -182,93 +389,465 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+            remote_value = 0;
+            success      = true;
+          }
+          else
+          {
+            log += F(" **no valid value**");
+          }
+          P021_remote[event->TaskIndex] = remote_value;
+        }
+        else
+        {
+          log += F(" **no valid command**");
+        }
+      }
+      P021_evaluate(event);
+      addLogMove(LOG_LEVEL_INFO, log);
+      break;
+    }
+
     case PLUGIN_TEN_PER_SECOND:
     {
-      // we're checking a var from another task, so calculate that basevar
-      taskIndex_t TaskIndex = P021_CHECK_TASK;
+      if (!P021_getConfigBit(P021_FLAGS, P021_SLOW_EVAL))
+      {
+        P021_evaluate(event);
+      }
+      success = true;
+      break;
+    }
+    case PLUGIN_ONCE_A_SECOND:
+    {
+      P021_check_autosafe(event); // This function relies on being called exactly once a second
 
-      if (!validTaskIndex(TaskIndex)) {
-        break;
+      if (P021_getConfigBit(P021_FLAGS, P021_SLOW_EVAL))
+      {
+        P021_evaluate(event);
       }
-      uint8_t BaseVarIndex = TaskIndex * VARS_PER_TASK + P021_CHECK_VALUE;
-      float   value        = UserVar[BaseVarIndex];
-      uint8_t state        = switchstate[event->TaskIndex];
-
-      // compare with threshold value
-      bool  isZero             = essentiallyZero(P021_TRIGGER_HYSTERESIS);
-      float valueLowThreshold  = P021_TRIGGER_LEVEL - (isZero ? 0.0f : (P021_TRIGGER_HYSTERESIS / 2.0f));
-      float valueHighThreshold = P021_TRIGGER_LEVEL + (isZero ? 1.0f : (P021_TRIGGER_HYSTERESIS / 2.0f)); // Include setvalue on
-                                                                                                          // 0-hysteresis
-
-      if (!definitelyGreaterThan(value, valueLowThreshold)) {
-        state = P021_INVERT_OUTPUT == 0 ? 1 : 0;
+      success = true;
+      break;
+    }
+  }
+
+  return success;
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Convert control state to string for debugging purposes
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+#ifdef PLUGIN_021_DEBUG
+String P021_printControlState(int state)
+{
+  switch (state)
+  {
+    case P021_STATE_IDLE:     return F("Idle"); break;
+    case P021_STATE_HEATING:  return F("Heating"); break;
+    case P021_STATE_EXTEND:   return F("Extend"); break;
+    case P021_STATE_FORCE:    return F("Force"); break;
+    default:                  return F("***ERROR***"); break;
+  }
+}
+
+#endif // ifdef PLUGIN_021_DEBUG
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Convert control state to string for debugging purposes
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+#ifdef PLUGIN_021_DEBUG
+String P021_printControlMode(int mode)
```

Comment:
You're returning a flash string (`F()` macro), but have it converted to a `String` because of the return type defined. Better declare this as a `const __FlashStringHelper *`